hyper-capi/examples/client.c

```diff
@@ -0,0 +1,289 @@
+		return ret;
+	}
+}
+
+static int connect_to(char *host, char *port) {
+	struct addrinfo hints;
+	memset(&hints, 0, sizeof(struct addrinfo));
+	hints.ai_family = AF_UNSPEC;
+	hints.ai_socktype = SOCK_STREAM;
+
+	struct addrinfo *result, *rp;
+	if (getaddrinfo(host, port, &hints, &result) != 0) {
+		printf("dns failed for %s\n", host);
+		return -1;
+	}
+
+	int sfd;
+	for (rp = result; rp != NULL; rp = rp->ai_next) {
+		sfd = socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol);
+		if (sfd == -1) {
+			continue;
+		}
+
+		if (connect(sfd, rp->ai_addr, rp->ai_addrlen) != -1) {
+			break;
+		} else {
+			close(sfd);
+		}
+	}
+
+	freeaddrinfo(result);
+
+	// no address succeeded
+	if (rp == NULL) {
+		printf("connect failed for %s\n", host);
+		return -1;
+	}
+
+	return sfd;
+}
+
+static hyper_iter_step print_each_header(void *userdata, hyper_str name, hyper_str value) {
+	printf("%.*s: %.*s\n", (int) name.len, name.buf, (int) value.len, value.buf);
+	return HYPER_IT_CONTINUE;
+}
+
+int main(int argc, char *argv[]) {
+	printf("connecting ...\n");
+
+	int fd = connect_to("httpbin.org", "80");
+	if (fd < 0) {
+		return 1;
+	}
+	printf("connected to httpbin.org\n");
+
+	if (fcntl(fd, F_SETFL, O_NONBLOCK) != 0) {
+		printf("failed to set socket to non-blocking\n");
+		return 1;
+	}
+
+	fd_set fds_read;
+	fd_set fds_write;
+	fd_set fds_excep;
+
+	struct conn_data *conn = malloc(sizeof(struct conn_data));
+
+	conn->fd = fd;
+	conn->read_waker = NULL;
+	conn->write_waker = NULL;
+
+
+	// Hookup the IO
+	hyper_io *io = hyper_io_new();
+	hyper_io_set_data(io, (void *)conn);
+	hyper_io_set_read(io, read_cb);
+	hyper_io_set_write(io, write_cb);
+
+	printf("http handshake ...\n");
+
+	// We need an executor generally to poll futures
+	hyper_executor *exec = hyper_executor_new();
+
+	// Prepare client options
+	hyper_clientconn_options *opts = hyper_clientconn_options_new();
+	hyper_clientconn_options_exec(opts, exec);
+
+	hyper_task *handshake = hyper_clientconn_handshake(io, opts);
+
+
+	// Let's wait for the handshake to finish...
+	hyper_executor_push(exec, handshake);
+
+	// We're going to cheat for the handshake, since we know HTTP/1 handshakes
+	// are immediately ready after the first poll.
+	hyper_task *task = hyper_executor_poll(exec);
+	if (hyper_task_type(task) != HYPER_TASK_CLIENTCONN) {
+		// ruh roh!
+		printf("task not a handshake ?!\n");
+		return 1;
+	}
+
+	printf("preparing http request ...\n");
+
+	hyper_clientconn *client = hyper_task_value(task);
+
+
+	// Prepare the request
+	hyper_request *req = hyper_request_new();
+	if (hyper_request_set_method(req, (uint8_t *)"GET", 3)) {
+		printf("error setting method\n");
+		return 1;
+	}
+	if (hyper_request_set_uri(req, (uint8_t *)"/", sizeof("/") - 1)) {
+		printf("error setting uri\n");
+		return 1;
+	}
+
+	hyper_headers *req_headers = hyper_request_headers(req);
+
+	hyper_str name = { .buf = "host", .len = 4};
+	hyper_str val = { .buf = "httpbin.org", .len = sizeof("httpbin.org") - 1 };
```

Comment:
These lines causes compiler warnings with `-pedantic` since `buf` is defined as unsigned while a const string is not:
~~~
examples/client.c: In function ‘main’:
examples/client.c:181:28: warning: pointer targets in initialization of ‘const uint8_t *’ {aka ‘const unsigned char *’} from ‘char *’ differ in signedness [-Wpointer-sign]
  181 |  hyper_str name = { .buf = "host", .len = 4};
      |                            ^~~~~~
examples/client.c:181:28: note: (near initialization for ‘name.buf’)
examples/client.c:182:27: warning: pointer targets in initialization of ‘const uint8_t *’ {aka ‘const unsigned char *’} from ‘char *’ differ in signedness [-Wpointer-sign]
  182 |  hyper_str val = { .buf = "httpbin.org", .len = sizeof("httpbin.org") - 1 };
      |                           ^~~~~~~~~~~~~
examples/client.c:182:27: note: (near initialization for ‘val.buf’)